cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/ArtifactHttpHandlerTestBase.java

```diff
@@ -0,0 +1,139 @@
+ * the License.
+ */
+package co.cask.cdap.internal.app.services.http.handlers;
+
+import co.cask.cdap.WordCountApp;
+import co.cask.cdap.api.artifact.ArtifactInfo;
+import co.cask.cdap.api.artifact.ArtifactScope;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.client.MetadataClient;
+import co.cask.cdap.client.config.ClientConfig;
+import co.cask.cdap.client.config.ConnectionConfig;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.discovery.EndpointStrategy;
+import co.cask.cdap.common.discovery.RandomEndpointStrategy;
+import co.cask.cdap.common.utils.DirUtils;
+import co.cask.cdap.gateway.handlers.ArtifactHttpHandler;
+import co.cask.cdap.internal.app.runtime.artifact.ArtifactRepository;
+import co.cask.cdap.internal.app.services.http.AppFabricTestBase;
+import co.cask.cdap.internal.io.SchemaTypeAdapter;
+import co.cask.cdap.proto.id.ArtifactId;
+import co.cask.cdap.proto.id.NamespaceId;
+import com.google.common.base.Charsets;
+import com.google.common.io.ByteStreams;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import org.apache.twill.discovery.Discoverable;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.apache.twill.discovery.ServiceDiscovered;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.net.HttpURLConnection;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.concurrent.TimeUnit;
+import java.util.jar.Manifest;
+import javax.ws.rs.HttpMethod;
+
+/**
+ * TestBase for {@link ArtifactHttpHandler}. Contains common setup and other common methods
+ */
+public abstract class ArtifactHttpHandlerTestBase extends AppFabricTestBase {
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+    .create();
+  private static String systemArtifactsDir;
+  static ArtifactRepository artifactRepository;
+  static MetadataClient metadataClient;
+  private static ClientConfig clientConfig;
+
+  @BeforeClass
+  public static void setup() {
+    artifactRepository = getInjector().getInstance(ArtifactRepository.class);
+    systemArtifactsDir = getInjector().getInstance(CConfiguration.class).get(Constants.AppFabric.SYSTEM_ARTIFACTS_DIR);
+    DiscoveryServiceClient discoveryClient = getInjector().getInstance(DiscoveryServiceClient.class);
+    ServiceDiscovered metadataHttpDiscovered = discoveryClient.discover(Constants.Service.METADATA_SERVICE);
+    EndpointStrategy endpointStrategy = new RandomEndpointStrategy(metadataHttpDiscovered);
+    Discoverable discoverable = endpointStrategy.pick(1, TimeUnit.SECONDS);
+    Assert.assertNotNull(discoverable);
+    String host = "127.0.0.1";
```

Comment:
Why? The host should be coming from the `discoverable`?

Reply:
Refactored to a base class code from existing code.
Fixed.